### DISCUSSION TITLE
Collectable class

### DISCUSSION THREAD

**Discussion by quasiquote1** [sparcians/map#262] [Q&A] [answered]
In CoreModel's Execute.hpp, I see that there's an example of a pipeline collector object:

sparta::collection::Collectable<ExampleInstPtr> collected_inst_;

I also see examples with the execution time of the instruction is being collected:

collected_inst_.collectWithDuration(ex_inst, exe_time);

However I'm not seeing where this data is being used in the final report stats.

Am I missing something here?  What's the intended usage of this class?

Separately, is there a good example of how to go about collecting histograms of different latencies (say for memory responses)?   The intent is to be able at end of simulation to display a histogram of all encountered latencies.  I've only seen  histograms being output for sparta::Queue objects but not for anything else.

Thanks!

**Comment by klingaard** [comment, marked as answer]
Good questions.  When building a sparta model, there are three major focus areas for development:

1. The mode's general construction (layout) area using [Simulation](https://sparcians.github.io/map/classsparta_1_1app_1_1Simulation.html), [Unit](https://sparcians.github.io/map/classsparta_1_1Unit.html), and [Factories](https://sparcians.github.io/map/classsparta_1_1ResourceFactoryBase.html).  See the [Skeletal Pipeline](https://sparcians.github.io/map/skeleton_example.html) model for an example
2. The [model's communication](https://sparcians.github.io/map/communication.html) to simulate timing (i.e. putting together a representation of the hardware you're modeling) area using [Ports](https://sparcians.github.io/map/modeling.html#ports_overview) and [Events](https://sparcians.github.io/map/modeling.html#events_overview).  See the [Skeletal Pipeline](https://sparcians.github.io/map/skeleton_example.html) model for an example
3. Data collection area for analysis including [Statistic APIs](https://sparcians.github.io/map/modeling.html#statistics_overview) and [Collection](https://sparcians.github.io/map/modeling.html#collection_overview)

When all three are in place (like in the [Core Example](https://sparcians.github.io/map/core_example.html)) the performance architect can run the model and collect data to help find performance related issues.  That data is collected using modeler-added instances of [Counter](https://sparcians.github.io/map/classsparta_1_1Counter.html), Histogram Counters (many examples [here's one](https://sparcians.github.io/map/classsparta_1_1HistogramBase.html)), etc as part of the [Statistic APIs](https://sparcians.github.io/map/modeling.html#statistics_overview) (see an example [here](https://github.com/sparcians/map/blob/8a95417a2a75a8231339424e42186f3ed35ae705/sparta/example/CoreModel/src/Execute.hpp#L100)).   These counters are automatically collected by the framework and provided in a report either at the end of simulation or even during simulation at phases the perf architect requests (like every _N_ instructions completed).  The command line `--report[-all] <args>` is usually used to collect this information.  See "Generating Reports" on the [Core Example](https://sparcians.github.io/map/core_example.html) page.

_Collectables_ and _Collection_ are different and are _not_ related to reporting nor statistics.  They are related to _data collection_ similar to a VCD/fsdb data dump in hardware.  By default the sparta modeling framework will not collect any signal-based information without the modeler explicitly adding collectors for that data.  This is to keep the performance of the simulation high and not clutter generated output.  When collectors are added and enabled, this data is collected on every cycle and dumped into a database.  This is typically called "generating a pipeout."   The command `-z <pipeout name>` will generate that information.   After the database is generated, the modeler will use a tool called Argos to view the data graphically.  See "Generating Pipeouts" on the [Core Example](https://sparcians.github.io/map/core_example.html) page.

**Comment by klingaard** [comment]
> Separately, is there a good example of how to go about collecting histograms of different latencies (say for memory responses)? The intent is to be able at end of simulation to display a histogram of all encountered latencies. I've only seen histograms being output for sparta::Queue objects but not for anything else.

There is a class called [`HistogramStandalone`](https://sparcians.github.io/map/classsparta_1_1HistogramStandalone.html) that might be a good start.  There's another one called [`BasicHistogram`](https://sparcians.github.io/map/classsparta_1_1BasicHistogram.html) that might be more inline.
